### PR TITLE
cmd/list: adding description search and glob by default

### PR DESCRIPTION
**Pull request by trws**
Generally making list a little more flexible.

### CODE REVIEW

File: lib/spack/spack/cmd/list.py [hunk 1]
```diff
@@ -39,26 +41,45 @@ def setup_parser(subparser):
     subparser.add_argument(
         '-i', '--insensitive', action='store_true', default=False,
         help='Filtering will be case insensitive.')
+    subparser.add_argument(
+        '-s', '--search_description', action='store_true', default=False,
```
Comment (tgamblin) [inline lib/spack/spack/cmd/list.py]: can this be `-d`?  In case I add `--search-something-else`?

Comment (tgamblin) [inline lib/spack/spack/cmd/list.py]: Also the long argument should be `--search-description`, not `--search_description`

Comment (trws) [inline lib/spack/spack/cmd/list.py]: Sure, I'll adjust it now.

**Comment by adamjstewart** [comment]
I really like the idea of globbing by default. That would've prevented my problems in #409 since Bash expands wildcards if they aren't in quotes. It would also prevent people from search for `tiff` and thinking they have to add a new package because it's actually called `libtiff`.

I don't see why the search shouldn't always be case-insensitive. Thoughts?

**Comment by tgamblin** [comment]
insensitive would be good

**Comment by trws** [comment]
My personal preference on that would be to have it be case insensitive *unless the user included a capitol letter in their query*.  That's how I have most of my search stuff set up, since it means you can search with case without specifying it with a flag but don't have to worry about it until you care.

**Comment by tgamblin** [comment]
I like it.  it's what emacs progressive search does :)

**Comment by adamjstewart** [comment]
Agreed! That's how vim's smartcase works as well.

**Comment by adamjstewart** [comment]
Also, can you update the documentation?

**Comment by trws** [comment]
Good point, I'll do it while I get the new behavior set.

**Comment by trws** [comment]
Ok, smartsearch-style case insensitive search and default globbing.  Also added some docs, especially examples of the implicit behavior.

**Comment by trws** [comment]
Hrm... Don't want to do this right now, but might be worth considering matching this behavior in find, if only for consistency.

File: lib/spack/docs/basic_usage.rst [hunk 2]
```diff
@@ -24,12 +24,29 @@ Spack can install:
 
 .. command-output:: spack list
 
-The packages are listed by name in alphabetical order.  You can also
-do wildcats searches using ``*``:
+The packages are listed by name in alphabetical order.  If you specify a
+pattern to match, it will follow this set of rules.  A pattern with no
+wildcards, ``*`` or ``?``, will be treated as though it started and ended iwth
+``*``, so ``util`` is equivalent to ``*util*``.  A pattern with no capitol
```
Comment (adamjstewart) [inline lib/spack/docs/basic_usage.rst]: *capital

File: lib/spack/docs/basic_usage.rst [hunk 3]
```diff
@@ -24,12 +24,29 @@ Spack can install:
...
 
-.. command-output:: spack list m*
+All packages whose names contain "sql" case insensitive:
 
-.. command-output:: spack list *util*
+.. command-output:: spack list sql
+
+All packages whose names start with a capitol M:
+
+.. command-output:: spack list M*
```
Comment (adamjstewart) [inline lib/spack/docs/basic_usage.rst]: As I learned in #409, this needs to have quotes around it.

File: lib/spack/docs/basic_usage.rst (hunk 2 above)
Comment (trws) [inline lib/spack/docs/basic_usage.rst]: Yup, had that typo twice. Thanks @adamjstewart.

**Comment by adamjstewart** [comment]
Do we still need to keep the `-i` option if it defaults to smartcase? I can't imagine very many people will want to type capital letters and search for lowercase letters.

File: lib/spack/docs/basic_usage.rst [hunk 4]
```diff
@@ -24,12 +24,29 @@ Spack can install:
 
 .. command-output:: spack list
 
-The packages are listed by name in alphabetical order.  You can also
-do wildcats searches using ``*``:
+The packages are listed by name in alphabetical order.  If you specify a
+pattern to match, it will follow this set of rules.  A pattern with no
+wildcards, ``*`` or ``?``, will be treated as though it started and ended iwth
```
Comment (adamjstewart) [inline lib/spack/docs/basic_usage.rst]: *with

Comment (trws) [inline lib/spack/docs/basic_usage.rst]: Yup, thanks.

**Comment by trws** [comment]
Perhaps it should be a "sensitive" option now rather than an "insensitive" option.  Not sure.

**Comment by adamjstewart** [comment]
Meh, it doesn't have to be perfect. The point is just to list packages that might be what you're looking for. No one is going to continuously refine their search until it gets down to a single package. I would be fine with it always being insensitive, not even smartcase. Especially for `--search-description`.

**Comment by trws** [comment]
Seems to be passing.  Any more to fix before this goes in?

**Comment by tgamblin** [comment]
Merged.  I changed the arg to `--case-sensitive`

**Pull request merged**